File: app/domain/preferences.py
"""Account-owned preferences with versioned updates; no global preference import."""
from typing import Literal
from fastapi import APIRouter,Request,HTTPException
from pydantic import BaseModel,ConfigDict,Field
from app.domain.common import Command,listing,put,execute,now
router=APIRouter(prefix='/api/v3/preferences',tags=['User preferences'])
class Values(BaseModel):
    model_config=ConfigDict(extra='forbid')
    taste:Literal['','清淡','咸香','酸甜','鲜香']=''
    spice:Literal['','不辣','微辣','中辣','较辣']=''
    duration:Literal['','under_30','30_to_60','any']=''
    dislikedIngredients:str=Field('',max_length=200)
class Criteria(BaseModel):
    model_config=ConfigDict(extra='forbid')
    budget:float|None=Field(None,gt=0,allow_inf_nan=False)
    difficulty_target:Literal['easy','medium','hard']|None=None
    max_calories_kcal:float|None=Field(None,gt=0,allow_inf_nan=False)
    min_protein_g:float|None=Field(None,gt=0,allow_inf_nan=False)
    max_fat_g:float|None=Field(None,gt=0,allow_inf_nan=False)
    max_carbohydrates_g:float|None=Field(None,gt=0,allow_inf_nan=False)
class Update(Command):
    expected_version:int=Field(ge=0,strict=True)
    values:Values|None=None
    recommendation:Criteria|None=None

def current(user):return next(iter(listing('user_preferences',user)),{'version':0,'values':Values().model_dump(),'recommendation':Criteria().model_dump()})
@router.get('')
def get(request:Request):return {'preferences':current(request.state.user_id)}
@router.put('')
def update(request:Request,body:Update):
    user=request.state.user_id
    def action():
        old=current(user)
        if old['version']!=body.expected_version:raise HTTPException(409,'偏好已在其他设备修改，请刷新后核对')
        return {'preferences':put('user_preferences',user,user,{'values':body.values.model_dump() if body.values is not None else old['values'],'recommendation':body.recommendation.model_dump() if body.recommendation is not None else old['recommendation'],'updated_at':now()},body.expected_version or None)}
    return execute(user,'preferences-save',body,action)

def scoring_preferences(user):
    import re
    values=current(user)['values']
    return {**{k:values[k] for k in ('taste','spice','duration') if values[k]},'disliked_ingredients':[v.strip() for v in re.split('[,，、\n]',values['dislikedIngredients']) if v.strip()]}

File: android-native/app/src/main/java/com/smartcooking/app/data/Preferences.kt
package com.smartcooking.app.data

import com.smartcooking.app.core.BusinessApi
import com.smartcooking.app.core.EmptyObject
import com.smartcooking.app.core.long
import com.smartcooking.app.core.obj
import kotlinx.serialization.json.JsonObject

/** Account preferences (`/api/v3/preferences`): taste values plus saved recommendation criteria. */
data class AccountPreferences(val version: Long, val values: JsonObject, val recommendation: JsonObject)

class PreferencesRepository(private val business: BusinessApi) {
    suspend fun load(): AccountPreferences {
        val p = business.get("/preferences").obj("preferences") ?: EmptyObject
        return AccountPreferences(p.long("version") ?: 0, p.obj("values") ?: EmptyObject, p.obj("recommendation") ?: EmptyObject)
    }
}

object PreferenceOptions {
    val tastes = listOf("清淡", "咸香", "酸甜", "鲜香")
    val spices = listOf("不辣", "微辣", "中辣", "较辣")
    val durations = listOf("under_30" to "30 分钟内", "30_to_60" to "30–60 分钟", "any" to "不限时长")
    val nutritionFields = listOf(
        "max_calories_kcal" to "每份热量上限（kcal）",
        "min_protein_g" to "每份蛋白质下限（g）",
        "max_fat_g" to "每份脂肪上限（g）",
        "max_carbohydrates_g" to "每份碳水上限（g）",
    )
    val difficulties = listOf("" to "不限制", "easy" to "简单", "medium" to "中等", "hard" to "较难")
}
